File: tests/node/assembler/patterns-assembler.test.ts
import { describe, it, expect, beforeEach, afterEach } from "vitest";
import * as fs from "node:fs";
import * as path from "node:path";
import { tmpdir } from "node:os";
import { PatternsAssembler } from "../../../src/assembler/patterns-assembler.js";
import { TemplateEngine } from "../../../src/template-engine.js";
import {
  ProjectConfig,
  ProjectIdentity,
  ArchitectureConfig,
  InterfaceConfig,
  LanguageConfig,
  FrameworkConfig,
} from "../../../src/models.js";

function buildConfig(overrides: {
  style?: string;
  eventDriven?: boolean;
} = {}): ProjectConfig {
  return new ProjectConfig(
    new ProjectIdentity("my-app", "A test application"),
    new ArchitectureConfig(
      overrides.style ?? "microservice",
      false,
      overrides.eventDriven ?? false,
    ),
    [new InterfaceConfig("rest")],
    new LanguageConfig("java", "21"),
    new FrameworkConfig("quarkus", "3.0", "maven"),
  );
}

function createPatternFile(
  resourcesDir: string,
  category: string,
  filename: string,
  content: string = `# ${filename}`,
): void {
  const dir = path.join(resourcesDir, "patterns", category);
  fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(path.join(dir, filename), content, "utf-8");
}

describe("PatternsAssembler", () => {
  let tmpDir: string;
  let resourcesDir: string;
  let outputDir: string;
  let assembler: PatternsAssembler;

  beforeEach(() => {
    tmpDir = fs.mkdtempSync(
      path.join(tmpdir(), "patterns-asm-test-"),
    );
    resourcesDir = path.join(tmpDir, "resources");
    fs.mkdirSync(resourcesDir, { recursive: true });
    outputDir = path.join(tmpDir, "output");
    fs.mkdirSync(outputDir, { recursive: true });
    assembler = new PatternsAssembler();
  });

  afterEach(() => {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  });

  describe("assemble", () => {
    it("returnsEmpty_whenUnknownArchitectureStyle", () => {
      const config = buildConfig({ style: "unknown-style" });
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      expect(result).toEqual([]);
    });

    it("returnsEmpty_whenNoPatternsOnDisk", () => {
      const config = buildConfig({ style: "microservice" });
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      expect(result).toEqual([]);
    });

    it("selectsUniversalPatterns_forLibraryStyle", () => {
      createPatternFile(resourcesDir, "architectural", "arch.md");
      createPatternFile(resourcesDir, "data", "data.md");
      const config = buildConfig({ style: "library" });
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      expect(result.length).toBeGreaterThanOrEqual(3);
      expect(result.some((f) => f.includes("architectural"))).toBe(
        true,
      );
      expect(result.some((f) => f.includes("data"))).toBe(true);
    });

    it.each([
      ["microservice", ["architectural", "data", "integration", "microservice", "resilience"]],
      ["hexagonal", ["architectural", "data", "integration"]],
      ["monolith", ["architectural", "data", "integration"]],
      ["library", ["architectural", "data"]],
    ])(
      "style_%s_includesCategories",
      (style, expectedCategories) => {
        for (const cat of expectedCategories) {
          createPatternFile(resourcesDir, cat, `${cat}-guide.md`);
        }
        const config = buildConfig({ style });
        const engine = new TemplateEngine(resourcesDir, config);
        const result = assembler.assemble(
          config, outputDir, resourcesDir, engine,
        );
        for (const cat of expectedCategories) {
          expect(
            result.some((f) => f.includes(cat)),
            `Expected category '${cat}' in results`,
          ).toBe(true);
        }
      },
    );

    it("includesEventDrivenPatterns_whenEnabled", () => {
      createPatternFile(resourcesDir, "architectural", "arch.md");
      createPatternFile(resourcesDir, "data", "data.md");
      createPatternFile(
        resourcesDir, "saga-pattern", "saga.md",
      );
      createPatternFile(
        resourcesDir, "outbox-pattern", "outbox.md",
      );
      createPatternFile(
        resourcesDir, "event-sourcing", "es.md",
      );
      createPatternFile(
        resourcesDir, "dead-letter-queue", "dlq.md",
      );
      const config = buildConfig({
        style: "microservice",
        eventDriven: true,
      });
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      expect(result.some((f) => f.includes("saga-pattern"))).toBe(
        true,
      );
      expect(
        result.some((f) => f.includes("outbox-pattern")),
      ).toBe(true);
    });

    it("excludesEventDrivenPatterns_whenDisabled", () => {
      createPatternFile(resourcesDir, "architectural", "arch.md");
      createPatternFile(
        resourcesDir, "saga-pattern", "saga.md",
      );
      const config = buildConfig({
        style: "microservice",
        eventDriven: false,
      });
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      expect(
        result.some((f) => f.includes("saga-pattern")),
      ).toBe(false);
    });

    it("writesFilesToReferencesWithCategoryStructure", () => {
      createPatternFile(
        resourcesDir, "architectural", "patterns.md",
      );
      createPatternFile(resourcesDir, "data", "data-guide.md");
      const config = buildConfig({ style: "library" });
      const engine = new TemplateEngine(resourcesDir, config);
      assembler.assemble(config, outputDir, resourcesDir, engine);
      const archFile = path.join(
        outputDir, "skills", "patterns", "references",
        "architectural", "patterns.md",
      );
      const dataFile = path.join(
        outputDir, "skills", "patterns", "references",
        "data", "data-guide.md",
      );
      expect(fs.existsSync(archFile)).toBe(true);
      expect(fs.existsSync(dataFile)).toBe(true);
    });

    it("createsConsolidatedSkillMd", () => {
      createPatternFile(
        resourcesDir, "architectural", "arch.md", "# Arch",
      );
      createPatternFile(
        resourcesDir, "data", "data.md", "# Data",
      );
      const config = buildConfig({ style: "library" });
      const engine = new TemplateEngine(resourcesDir, config);
      assembler.assemble(config, outputDir, resourcesDir, engine);
      const skillMd = path.join(
        outputDir, "skills", "patterns", "SKILL.md",
      );
      expect(fs.existsSync(skillMd)).toBe(true);
      const content = fs.readFileSync(skillMd, "utf-8");
      expect(content).toContain("# Arch");
      expect(content).toContain("# Data");
      expect(content).toContain("\n\n---\n\n");
    });

    it("consolidatedFileJoinsWithSeparator", () => {
      createPatternFile(
        resourcesDir, "architectural", "a.md", "AAA",
      );
      createPatternFile(
        resourcesDir, "data", "b.md", "BBB",
      );
      const config = buildConfig({ style: "library" });
      const engine = new TemplateEngine(resourcesDir, config);
      assembler.assemble(config, outputDir, resourcesDir, engine);
      const content = fs.readFileSync(
        path.join(outputDir, "skills", "patterns", "SKILL.md"),
        "utf-8",
      );
      expect(content).toBe("AAA\n\n---\n\nBBB");
    });

    it("singleFile_noSeparatorInConsolidated", () => {
      createPatternFile(
        resourcesDir, "architectural", "only.md", "ONLY",
      );
      const config = buildConfig({ style: "library" });
      const engine = new TemplateEngine(resourcesDir, config);
      assembler.assemble(config, outputDir, resourcesDir, engine);
      const content = fs.readFileSync(
        path.join(outputDir, "skills", "patterns", "SKILL.md"),
        "utf-8",
      );
      expect(content).toBe("ONLY");
      expect(content).not.toContain("---");
    });

    it("replacesPlaceholdersViaEngine", () => {
      createPatternFile(
        resourcesDir,
        "architectural",
        "arch.md",
        "Project: {project_name}",
      );
      const config = buildConfig({ style: "library" });
      const engine = new TemplateEngine(resourcesDir, config);
      assembler.assemble(config, outputDir, resourcesDir, engine);
      const refContent = fs.readFileSync(
        path.join(
          outputDir, "skills", "patterns", "references",
          "architectural", "arch.md",
        ),
        "utf-8",
      );
      expect(refContent).toBe("Project: my-app");
      const skillContent = fs.readFileSync(
        path.join(outputDir, "skills", "patterns", "SKILL.md"),
        "utf-8",
      );
      expect(skillContent).toBe("Project: my-app");
    });

    it("returnsAllWrittenFilePaths", () => {
      createPatternFile(
        resourcesDir, "architectural", "a.md",
      );
      createPatternFile(resourcesDir, "data", "b.md");
      const config = buildConfig({ style: "library" });
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      expect(result).toHaveLength(3);
      const refFiles = result.filter((f) =>
        f.includes("references"),
      );
      const skillFile = result.filter((f) =>
        f.endsWith("SKILL.md"),
      );
      expect(refFiles).toHaveLength(2);
      expect(skillFile).toHaveLength(1);
    });

    it("multipleFilesPerCategory_allWritten", () => {
      createPatternFile(
        resourcesDir, "architectural", "a.md", "A",
      );
      createPatternFile(
        resourcesDir, "architectural", "b.md", "B",
      );
      const config = buildConfig({ style: "library" });
      const engine = new TemplateEngine(resourcesDir, config);
      const result = assembler.assemble(
        config, outputDir, resourcesDir, engine,
      );
      const archFiles = result.filter((f) =>
        f.includes(path.join("references", "architectural")),
      );
      expect(archFiles).toHaveLength(2);
    });
  });
});

File: java/src/test/java/dev/iadev/assembler/PatternsAssemblerTest.java
package dev.iadev.assembler;

import dev.iadev.domain.model.ProjectConfig;
import dev.iadev.template.TemplateEngine;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Tests for PatternsAssembler — the fourth assembler in
 * the pipeline, generating .claude/skills/patterns/
 * with consolidated SKILL.md and per-category reference
 * files.
 */
@DisplayName("PatternsAssembler")
class PatternsAssemblerTest {

    @Nested
    @DisplayName("assemble — implements Assembler interface")
    class ImplementsAssembler {

        @Test
        @DisplayName("is instance of Assembler")
        void instanceOf_whenCreated_implementsAssemblerInterface() {
            PatternsAssembler assembler =
                    new PatternsAssembler();

            assertThat(assembler)
                    .isInstanceOf(Assembler.class);
        }
    }

    @Nested
    @DisplayName("assemble — microservice architecture"
            + " includes all categories")
    class MicroserviceArchitecture {

        @Test
        @DisplayName("microservice generates architectural"
                + " pattern references")
        void assemble_microservice_includesArchitectural(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config = buildMicroserviceConfig();

            assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            Path refsDir = outputDir.resolve(
                    "skills/patterns/references");
            assertThat(refsDir.resolve(
                    "architectural/cqrs.md"))
                    .exists();
            assertThat(refsDir.resolve(
                    "architectural/hexagonal-architecture.md"))
                    .exists();
        }

        @Test
        @DisplayName("microservice generates microservice"
                + " pattern references")
        void assemble_microservice_includesMicroservice(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config = buildMicroserviceConfig();

            assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            Path refsDir = outputDir.resolve(
                    "skills/patterns/references");
            assertThat(refsDir.resolve(
                    "microservice/saga-pattern.md"))
                    .exists();
            assertThat(refsDir.resolve(
                    "microservice/api-gateway.md"))
                    .exists();
        }

        @Test
        @DisplayName("microservice generates resilience"
                + " pattern references")
        void assemble_microservice_includesResilience(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config = buildMicroserviceConfig();

            assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            Path refsDir = outputDir.resolve(
                    "skills/patterns/references");
            assertThat(refsDir.resolve(
                    "resilience/circuit-breaker.md"))
                    .exists();
            assertThat(refsDir.resolve(
                    "resilience/retry-with-backoff.md"))
                    .exists();
        }

        @Test
        @DisplayName("microservice generates integration"
                + " pattern references")
        void assemble_microservice_includesIntegration(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config = buildMicroserviceConfig();

            assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            Path refsDir = outputDir.resolve(
                    "skills/patterns/references");
            assertThat(refsDir.resolve(
                    "integration/adapter-pattern.md"))
                    .exists();
        }

        @Test
        @DisplayName("microservice generates data"
                + " pattern references")
        void assemble_microservice_includesData(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config = buildMicroserviceConfig();

            assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            Path refsDir = outputDir.resolve(
                    "skills/patterns/references");
            assertThat(refsDir.resolve(
                    "data/repository-pattern.md"))
                    .exists();
        }

        @Test
        @DisplayName("microservice generates consolidated"
                + " SKILL.md")
        void assemble_microservice_generatesConsolidated(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config = buildMicroserviceConfig();

            assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            Path skillMd = outputDir.resolve(
                    "skills/patterns/SKILL.md");
            assertThat(skillMd).exists();
            String content = Files.readString(
                    skillMd, StandardCharsets.UTF_8);
            assertThat(content).contains("---");
        }

        @Test
        @DisplayName("returned list includes all generated"
                + " files")
        void assemble_returnedList_includesAll(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config = buildMicroserviceConfig();

            List<String> files = assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            // 22 pattern files + 1 consolidated SKILL.md
            assertThat(files).hasSizeGreaterThan(10);
            assertThat(files).anyMatch(
                    f -> f.endsWith("SKILL.md"));
        }
    }

    @Nested
    @DisplayName("assemble — library architecture"
            + " excludes microservice patterns")
    class LibraryArchitecture {

        @Test
        @DisplayName("library includes only architectural"
                + " and data patterns")
        void assemble_library_includesOnlyArchAndData(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config =
                    TestConfigBuilder.builder()
                            .archStyle("library")
                            .eventDriven(false)
                            .container("none")
                            .orchestrator("none")
                            .iac("none")
                            .clearInterfaces()
                            .addInterface("cli")
                            .build();

            assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            Path refsDir = outputDir.resolve(
                    "skills/patterns/references");
            // architectural and data are present
            assertThat(refsDir.resolve(
                    "architectural")).isDirectory();
            assertThat(refsDir.resolve(
                    "data")).isDirectory();
            // microservice, resilience, integration absent
            assertThat(refsDir.resolve(
                    "microservice")).doesNotExist();
            assertThat(refsDir.resolve(
                    "resilience")).doesNotExist();
            assertThat(refsDir.resolve(
                    "integration")).doesNotExist();
        }

        @Test
        @DisplayName("library SKILL.md does not contain"
                + " microservice-only pattern content")
        void assemble_librarySkillMd_excludesMicroservice(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config =
                    TestConfigBuilder.builder()
                            .archStyle("library")
                            .eventDriven(false)
                            .container("none")
                            .orchestrator("none")
                            .iac("none")
                            .clearInterfaces()
                            .addInterface("cli")
                            .build();

            assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            Path skillMd = outputDir.resolve(
                    "skills/patterns/SKILL.md");
            assertThat(skillMd).exists();
            String content = Files.readString(
                    skillMd, StandardCharsets.UTF_8);
            // Saga Pattern is a microservice-specific file
            assertThat(content)
                    .doesNotContain("# Saga Pattern");
            // Bulkhead is a microservice-specific file
            assertThat(content)
                    .doesNotContain("# Bulkhead");
        }
    }

    @Nested
    @DisplayName("assemble — unknown architecture style")
    class UnknownArchitecture {

        @Test
        @DisplayName("unknown style returns empty list")
        void assemble_unknownStyle_returnsEmpty(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config =
                    TestConfigBuilder.builder()
                            .archStyle("exotic-style")
                            .eventDriven(false)
                            .container("none")
                            .orchestrator("none")
                            .iac("none")
                            .clearInterfaces()
                            .addInterface("cli")
                            .build();

            List<String> files = assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            assertThat(files).isEmpty();
        }
    }

    @Nested
    @DisplayName("assemble — golden file parity")
    class GoldenFile {

        @Test
        @DisplayName("consolidated SKILL.md matches"
                + " golden file for go-gin profile")
        void assemble_skillMd_matchesGolden(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config = buildGoGinConfig();

            assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            String goldenPath =
                    "golden/go-gin/.claude/skills/patterns/"
                            + "SKILL.md";
            String expected = loadResource(goldenPath);
            if (expected != null) {
                String actual = Files.readString(
                        outputDir.resolve(
                                "skills/patterns/SKILL.md"),
                        StandardCharsets.UTF_8);
                assertThat(actual)
                        .as("patterns/SKILL.md")
                        .isEqualTo(expected);
            }
        }

        @Test
        @DisplayName("all pattern reference files match"
                + " golden files for go-gin profile")
        void assemble_allRefsMatchGolden_succeeds(
                @TempDir Path tempDir)
                throws IOException {
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler();
            ProjectConfig config = buildGoGinConfig();

            assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            String[][] refFiles = {
                    {"architectural", "cqrs.md"},
                    {"architectural", "event-sourcing.md"},
                    {"architectural",
                            "hexagonal-architecture.md"},
                    {"architectural", "modular-monolith.md"},
                    {"data", "cache-aside.md"},
                    {"data", "event-store.md"},
                    {"data", "repository-pattern.md"},
                    {"data", "unit-of-work.md"},
                    {"integration", "adapter-pattern.md"},
                    {"integration",
                            "anti-corruption-layer.md"},
                    {"integration",
                            "backend-for-frontend.md"},
                    {"microservice", "api-gateway.md"},
                    {"microservice", "bulkhead.md"},
                    {"microservice", "idempotency.md"},
                    {"microservice", "outbox-pattern.md"},
                    {"microservice", "saga-pattern.md"},
                    {"microservice", "service-discovery.md"},
                    {"microservice", "strangler-fig.md"},
                    {"resilience", "circuit-breaker.md"},
                    {"resilience", "dead-letter-queue.md"},
                    {"resilience", "retry-with-backoff.md"},
                    {"resilience", "timeout-patterns.md"},
            };

            for (String[] pair : refFiles) {
                String relPath = pair[0] + "/" + pair[1];
                String goldenPath =
                        "golden/go-gin/.claude/skills/"
                                + "patterns/references/"
                                + relPath;
                String expected =
                        loadResource(goldenPath);
                if (expected != null) {
                    String actual = Files.readString(
                            outputDir.resolve(
                                    "skills/patterns/"
                                            + "references/"
                                            + relPath),
                            StandardCharsets.UTF_8);
                    assertThat(actual)
                            .as("Pattern ref: " + relPath)
                            .isEqualTo(expected);
                }
            }
        }

        private String loadResource(String path) {
            var url = getClass().getClassLoader()
                    .getResource(path);
            if (url == null) {
                return null;
            }
            try {
                return Files.readString(
                        Path.of(url.getPath()),
                        StandardCharsets.UTF_8);
            } catch (IOException e) {
                return null;
            }
        }
    }

    @Nested
    @DisplayName("assemble — edge cases")
    class EdgeCases {

        @Test
        @DisplayName("custom resourcesDir with no"
                + " patterns dir returns empty")
        void assemble_emptyResources_returnsEmpty(
                @TempDir Path tempDir)
                throws IOException {
            Path resourceDir = tempDir.resolve("res");
            Files.createDirectories(resourceDir);
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler(resourceDir);
            ProjectConfig config =
                    TestConfigBuilder.minimal();

            List<String> files = assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            assertThat(files).isEmpty();
        }

        @Test
        @DisplayName("patterns dir exists but category"
                + " dir missing returns empty")
        void assemble_missingCategoryDir_returnsEmpty(
                @TempDir Path tempDir)
                throws IOException {
            Path resourceDir = tempDir.resolve("res");
            Path patternsDir =
                    resourceDir.resolve("patterns");
            Files.createDirectories(patternsDir);
            Path outputDir = tempDir.resolve("output");
            Files.createDirectories(outputDir);

            PatternsAssembler assembler =
                    new PatternsAssembler(resourceDir);
            ProjectConfig config =
                    TestConfigBuilder.builder()
                            .archStyle("library")
                            .eventDriven(false)
                            .build();

            List<String> files = assembler.assemble(
                    config, new TemplateEngine(), outputDir);

            assertThat(files).isEmpty();
        }
    }

    private static ProjectConfig buildMicroserviceConfig() {
        return TestConfigBuilder.builder()
                .archStyle("microservice")
                .eventDriven(true)
                .build();
    }

    private static ProjectConfig buildGoGinConfig() {
        var builder = TestConfigBuilder.builder()
                .projectName("my-go-service")
                .purpose(
                        "Describe your service purpose here")
                .archStyle("microservice")
                .domainDriven(false)
                .eventDriven(true)
                .language("go", "1.22")
                .framework("gin", "")
                .buildTool("go-mod")
                .nativeBuild(false);
        return configureGoInfra(builder).build();
    }

    private static TestConfigBuilder configureGoInfra(
            TestConfigBuilder builder) {
        return builder
                .container("docker")
                .orchestrator("kubernetes")
                .iac("terraform")
                .apiGateway("kong")
                .smokeTests(true)
                .contractTests(false)
                .performanceTests(true)
                .clearInterfaces()
                .addInterface("rest")
                .addInterface("grpc")
                .addInterface("event-consumer",
                        "", "kafka")
                .addInterface("event-producer",
                        "", "kafka");
    }
}
